# TestDrive assignment: Like the real assignment, but optional and does not count towards your grade.

In this notebook:

- explanation of the assignment
- marking criteria
- data and starting code for your 'testdrive' assignment
- submission template

# What is TestDrive this assignment about:

In this assignment you will role-play (pretend) that you are a junior data analyst for NHS.

Your task is to look at a dataset of all 15060 GP practices in England and Wales. What can you tell about this set? Are there any insights about the locations or activity of these in there, even thou the data is quite limited? Try to find something interesting, then visualise it and describe what you found in 200 words.

# What to do for this TrestDrive assignment?

- Have a look at the data, try writing some code that tells you what information is there, and what could be interesting to look at?
- identify an interesting question that this data could answer (see examples below)
- write code that will answer that question. If writing helper functions, can they be tested?
- visualise your findings as a simple graph, or even a printed text (try to lay it out so that it looks good).
- write abotu 2 paragraphs of text (200 words +/- 10%) explaining what you saw in the data, what question you identified, and how you answered it with code.

# How it fits within the course? How to submit and how to get feedback?

- if you can come to the pair programming lab for week 4, bring your work with you. During the lab you will be able to discuss it with your partner and with the instructors.
- to submit the TestDrive assignment, submit your code, and report via Learn in the Forum. Use the code button {...} when including python code, if you can, so that it is better formatted. PLEASE INCLUDE YOUR NOTEBOOK AS AN ATTACHMENT if you know how.
- all those who submit their TestDrive assignment until the date mentioned in Learn, will recieve feedback. Everybody's feedback will be visible to everybody, so that we can all learn better. (Remember this feedback does nto have impact on your grades and is just for you to learn better)

## Differences and simmilarities between this TestDrive assignment and the final actual assignment:

Simmilarities:


- identical sumbission components (code, visualisation, mini-report)
- simmilar data and tasks.
- you will be given feedback and points

Differences:

- points do not count towards your mark (this testdrive assignment is purely 'formative' i.e. here to help you learn more and better)
- you will submit this via Learn Forum, and will be given feedback there. This way everyone can learn from everyone else's code and feedback.
- here you will only ask and answer ONE task i.e. ONE 'business question'. So you are writing 1 mini-report. 

## 1. Deliverables FOR EACH TASK: Code + Mini-report + Visualisation

(in this TestDrive you have just one Task, btu final assignment will have 2 separate tasks)

For each task, please include the following:

- all the **code** that you used to arrive at the conclusion. 
- a cell with your **mini-report** (200 words +/- 10%). (example in notebook `Mini-report example`)
- a cell that will generate the **visualisation** (graph, table or print) that accompanies your mini-report. (read how to manipulate strings in `badge 9` and how to create graphs in `badge 12`)

### Code

In your notebook please include all the **code** that you used to arrive at the conclusion. 

- It is absolutely ok to re-use your own code from one question in another question
- But please write the most important parts of code for each Question that help you to answer the question 


### Mini-report

At the end of your notebook, for each Task, you should provide:

- a markdown (text) cell with your mini-report (200 words +/- 10%). For reference, 'Marking criteria' section below has around 450 words)

### Visualisation

Use the data you extracted to further your argument with visualisations:

- a cell that will generate the graph or table that accompanies your mini-report. You are allowed to combine a number of graphs (eg. three similar graphs side-by-side) or prints, if they work together and advance your argument.

## 2. Marking Criteria for each Mini Report: 

**Business Question and Answer:**

Did you manage to find a question that can be answered with given data? As an analyst, you will often have to dive into available data and identify how it can help the business, or solve a problem, without first knowing what the problem is. Also, can you formulate a clear answer to the question you created?
 
- 50% - C - GOOD - question and answer are clear, well defined and connect with the dataset
- 60% - B - VERY GOOD - argument is clearly positioned in a business context, and attempts to provide value/insight
- 70% - A - EXCELLENT - insights are novel, actionable and the writeup is of publishable quality.

**Using the Data:**

How well did you use the data to answer your question? Your answer should be supported by what you found in the data. Briefly describe why this was the correct data, and the correct analysis to perform on it.
 
- 50% - C - GOOD - data selected is appropriate for the task, analysis is clear, the source is mentioned
- 60% - B - VERY GOOD - data analysis section advances the argument, makes a clear point and is easy to read and understand
- 70% - A - EXCELLENT - analysis is insightful, using multiple parts of the data set in a creative way

**Visualisation:**

Can you aid your argument/answer with visual clues? A graph can say a thousand words, but it is also easy to make one that is confusing, or misleading. Use simple (or highly customised) graphs to make your argument clearer.
 
- 50% - C - GOOD - graph is communicative, appropriate and similar complexity as in the notes
- 60% - B - VERY GOOD - graph is customised and combines a number of styles and types of visualisation
- 70% - A - EXCELLENT - graph is using clear visual language to make a point, adds to the argument, and is of publishable quality

**Code Quality:**

Is your code clean, readable and DRY (Don't repeat yourself)? Are you using good readable variable names? Did you clean up your code and does it not include any old/unused parts?
 
- 50% - C - GOOD - code has meaningful variable names, no needlessly repeated code
- 60% - B - VERY GOOD - also signposted, reasonably commented and cleaned up
- 70% - A - EXCELLENT - also code has a logical flow, consistency of names and granularity/size

**Code Structure:**

Is your code well structured and broken down? Just like good writing has sentences, paragraphs and chapters, good code should be split into sections. Break down your code into cells and functions. Use meaningful signposts (eg. comments, function names) to guide the reader through your code.
 
- 50% - C - GOOD - code is broken down into cells, by the code's purpose
- 60% - B - VERY GOOD - code is broken down by cell and also separated and readable. Attempts on reusing code are made
- 70% - A - EXCELLENT - code is split into functions and/or objects and can be easily reused

## 3. Dataset and Example Questions:

You are given a dataset with information about 15060 practices in England and Wales. Have a look at this data. What can you say about it? What is interesting in this data, espacially for the NHS coordinators: what cities have most practices? Which ones are still operational? How long do practices exist (if they have 'Close Date')?

Do not try to find something complicated, just provide some overview of the data. Also there is no need to analyse data (corelations etc.). It's better to try to describe and visualise what you see in the data.

Example data about one GP practice:

```
{'Adress': {'Address line 1': 'THE HEALTH CENTRE',
            'Address line 2': 'LAWSON STREET',
            'Address line 3': 'STOCKTON-ON-TEES',
            'Address line 4': 'CLEVELAND',
            'Area': 'CLEVELAND',
            'City': 'STOCKTON-ON-TEES',
            'Full Postal Address': 'THE HEALTH CENTRE, LAWSON STREET, '
                                   'STOCKTON-ON-TEES, CLEVELAND',
            'Postcode': 'TS18 1HU',
            'Telephone': '01642 672351'},
 'Name': 'THE DENSHAM SURGERY',
 'Organisation Code': 'A81001',
 'Prescribing Setting': '4',
 'Status': {'Close Date': '', 'Open Date': '19740401', 'Status Code': 'A'}}
 ```

### Example questions:

- what are the cities where there are most practices? Can you visualise the most popular ones?
- Look at how many practices are 'Closed' and how many are 'Active'? Are there differences between these for large cities (those with many GPs)?
- For 'Closed' practices, how long did they exist on average?

Use one of these questions, or come up with your own. Make sure you clearly state your question in the mini-report.

### Starting code: No need to change it. This code just loads data for you:

In [12]:
# this is commented out, because it should not be needed on noteable 
# (but might be needed if you are using notebooks on your own laptop, with anaconda)
# if it does not work (on your own machine), you might also try the command below with pip3 instead of pip

# !pip install plotly --upgrade
# !pip3 install plotly --upgrade

In [13]:
# the usual imports
from plotly.subplots import make_subplots
import plotly.graph_objects as go

In [1]:
import pprint as pp
from datetime import datetime, timedelta, date, timezone
import json

In [15]:
# Let's load the data from the files.
# original data from:
# gp data from https://digital.nhs.uk/services/organisation-data-service/file-downloads/gp-and-gp-practice-related-data

In [5]:
# this function just loads the data from files, there is no need to understand how it does it.
def load_json_file_named(file_name):
    try: 
        loaded_data = []
        file_location = f"./data/{file_name}"
        with open(file_location, 'r') as file: # or f"data/{file_name}" depending on your files
            loaded_data =  json.load(file)
    except OSError as e:
        print(f"Error. Does the file exist in this folder? {file_location}\n\n {e}")
    return loaded_data

# btw. above we use here error try-catching:
# try is like "while there are no trouble, continue". If there is trouble... jump to 'except'
# see if work: change folder name above from 'data' to something silly, like 'banana'.
# you will see error: "No such file or directory"

In [6]:
gp_practices = load_json_file_named('gp_practices.json')
print(f'all files loaded. Data length: {len(gp_practices)}')
print("first data point:")
pp.pprint(gp_practices[0])

all files loaded. Data length: 15060
first data point:
{'Address': {'Address line 1': 'THE HEALTH CENTRE',
             'Address line 2': 'LAWSON STREET',
             'Address line 3': 'STOCKTON-ON-TEES',
             'Address line 4': 'CLEVELAND',
             'Area': 'CLEVELAND',
             'City': 'STOCKTON-ON-TEES',
             'Full Postal Address': 'THE HEALTH CENTRE, LAWSON STREET, '
                                    'STOCKTON-ON-TEES, CLEVELAND',
             'Postcode': 'TS18 1HU',
             'Telephone': '01642 672351'},
 'Name': 'THE DENSHAM SURGERY',
 'Organisation Code': 'A81001',
 'Prescribing Setting': '4',
 'Status': {'Close Date': '', 'Open Date': '19740401', 'Status Code': 'A'}}


### End of starting code!

## How to interpret this data?

you now have a variable `gp_practices` which holds information about 15060 GP practoces in England (and a few around UK).

Look at this data. Most of it is self explanatory. It has 5 keys: 'Organisation Code', 'Name', 'Adress', 'Status', 'Prescribing Setting'. Some keys, are more complex, like Address or Status, and hold information inside. 


Some values are encoded, eg. Status Code can hold a letter A, C, D or P. See below for the key.

```
A = Active
C = Closed
D = Dormant 
P = Proposed
```

While Prescribing Setting, hold a number, see key below:

```
0 = Other
1 = WIC Practice
2 = OOH Practice
3 = WIC + OOH Practice
4 = GP Practice
8 = Public Health Service
9 = Community Health Service
10 = Hospital Service
11 = Optometry Service
12 = Urgent & Emergency Care
13 = Hospice
14 = Care Home / Nursing Home
15 = Border Force
16 = Young Offender Institution
17 = Secure Training Centre
18 = Secure Children's Home
19 = Immigration Removal Centre 20 = Court
21 = Police Custody
22 = Sexual Assault Referral Centre (SARC)
24 = Other – Justice Estate
25 = Prison
```

In [18]:
# let's grab one GP practice:
pp.pprint(gp_practices[0])

{'Address': {'Address line 1': 'THE HEALTH CENTRE',
             'Address line 2': 'LAWSON STREET',
             'Address line 3': 'STOCKTON-ON-TEES',
             'Address line 4': 'CLEVELAND',
             'Area': 'CLEVELAND',
             'City': 'STOCKTON-ON-TEES',
             'Full Postal Address': 'THE HEALTH CENTRE, LAWSON STREET, '
                                    'STOCKTON-ON-TEES, CLEVELAND',
             'Postcode': 'TS18 1HU',
             'Telephone': '01642 672351'},
 'Name': 'THE DENSHAM SURGERY',
 'Organisation Code': 'A81001',
 'Prescribing Setting': '4',
 'Status': {'Close Date': '', 'Open Date': '19740401', 'Status Code': 'A'}}


In [7]:
[gp['Address']["Area"]
    for gp in gp_practices
    if gp['Prescribing Setting'] == '4'
]

['CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'GUISBOROUGH',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'MIDDLESBROUGH',
 'SALTBURN BY SEA',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'YARM, CLEVELAND',
 'MIDDLESBROUGH',
 'MIDDLESBROUGH',
 'CLEVELAND',
 'CLEVELAND',
 'MIDDLESBROUGH',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'MIDDLESBROUGH',
 'CLEVELAND',
 'STOCKTON-ON-TEES',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'REDCAR,CLEVELAND',
 'CLEVELAND',
 'MIDDLESBROUGH',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'CLEVELAND',
 'MIDDLESBROUGH',
 'CLEVELAND',
 'HARTLEPOOL',
 'MIDDLESBROUGH',
 'CLEVELAND',
 'CLEVELAND',
 'STOCKTON-ON-TEES',
 'CLEVELAND',
 'HARTLEPOOL',
 'CLEVELAND',
 'CLEVELAND',
 'STOCKTON ON SEA',
 'CLEVELAND',
 'CLEVELAND',
 

In [20]:
# It has 5 keys: 'Organisation Code', 'Name', 'Adress', 'Status', 'Prescribing Setting'
pp.pprint(gp_practices[0].keys())

dict_keys(['Organisation Code', 'Name', 'Address', 'Status', 'Prescribing Setting'])


### Example code: how many 'Prescribing Setting' of each type there are?

In [21]:
# To get more info about what data means have a look at the file gppractices.pdf in /data folder
# or google around it. 
# note that the data provider does not really explain what 'Prescribing Setting' means :(
# So we need to guess, or figure it out.
def prescribing_setting_report(all_practices):
    prescribing_settings_meanings = {
        '0':'Other',
        '1':'WIC Practice',
        '2':'OOH Practice',
        '3':'WIC + OOH Practice',
        '4':'GP Practice',
        '8':'Public Health Service',
        '9':'Community Health Service',
        '10':'Hospital Service',
        '11':'Optometry Service',
        '12':'Urgent & Emergency Care',
        '13':'Hospice',
        '14':'Care Home / Nursing Home',
        '15':'Border Force',
        '16':'Young Offender Institution',
        '17':'Secure Training Centre',
        '18':'Secure Children/s Home',
        '19':'Immigration Removal Centre',
        '20':'Court',
        '21':'Police Custody',
        '22':'Sexual Assault Referral Centre (SARC)',
        '24':'Other – Justice Estate',
        '25':'Prison'}

    all_settings = [gp['Prescribing Setting'] 
                for gp in all_practices ]
    
    settings_unique = set(all_settings)

    for setting in settings_unique:
        meanings = prescribing_settings_meanings[setting]
        frequency = all_settings.count(setting)
        print(f"{meanings:.<40} occures {frequency: >5} times")

prescribing_setting_report(gp_practices)

Community Health Service................ occures  1734 times
GP Practice............................. occures  8632 times
Young Offender Institution.............. occures     5 times
Secure Training Centre.................. occures     2 times
Other................................... occures  2557 times
WIC Practice............................ occures   121 times
Optometry Service....................... occures   120 times
Urgent & Emergency Care................. occures   151 times
Prison.................................. occures   149 times
Hospital Service........................ occures   578 times
Public Health Service................... occures   356 times
WIC + OOH Practice...................... occures    22 times
Care Home / Nursing Home................ occures    91 times
Immigration Removal Centre.............. occures     9 times
OOH Practice............................ occures   357 times
Hospice................................. occures   154 times
Sexual Assault Referral 

# TestDrive Assignment Solution:

## Your Code:
For 'Closed' practices, how long did they exist on average?
"""
{'Address': {'Address line 1': 'THE HEALTH CENTRE',
             'Address line 2': 'LAWSON STREET',
             'Address line 3': 'STOCKTON-ON-TEES',
             'Address line 4': 'CLEVELAND',
             'Area': 'CLEVELAND',
             'City': 'STOCKTON-ON-TEES',
             'Full Postal Address': 'THE HEALTH CENTRE, LAWSON STREET, '
                                    'STOCKTON-ON-TEES, CLEVELAND',
             'Postcode': 'TS18 1HU',
             'Telephone': '01642 672351'},
 'Name': 'THE DENSHAM SURGERY',
 'Organisation Code': 'A81001',
 'Prescribing Setting': '4',
 'Status': {'Close Date': '', 'Open Date': '19740401', 'Status Code': 'A'}}
 """

In [8]:
def get_opening_closing(data):
    closed_practices = [
        {"Closed":practice["Status"]["Close Date"],"Opened": practice["Status"]["Open Date"]}
        for practice in data
        if practice["Status"]["Status Code"] == "C"
    ]
    return closed_practices

closed_practices_list = get_opening_closing(gp_practices)
print(closed_practices_list[0])

{'Closed': '20171031', 'Opened': '19740401'}


In [9]:
from datetime import datetime, timedelta, date, timezone

def convert_string_to_date(date_string):
    date_object = datetime.strptime(date_string, "%Y%m%d")
    return date_object

closed_practices_dates_list = [
    {'Closed Date': convert_string_to_date(date['Closed']), 'Opened Date': convert_string_to_date(date['Opened'])}
    for date in closed_practices_list
]

print(closed_practices_dates_list[0])

{'Closed Date': datetime.datetime(2017, 10, 31, 0, 0), 'Opened Date': datetime.datetime(1974, 4, 1, 0, 0)}


In [10]:
import statistics as stat

opened_days = 0
practices_list = []
for i in range(len(closed_practices_dates_list)):
    timedelta = closed_practices_dates_list[i]["Closed Date"] - closed_practices_dates_list[i]["Opened Date"]
    opened_days+=int(timedelta.days)
    practices_list.append(int(timedelta.days))
    
mean_days = opened_days/len(closed_practices_dates_list)
print(mean_days)
median_days = stat.median(practices_list)
print(median_days)

8269.736025848142
6421


## Visualisation:

In [11]:
import plotly.express as px

fig = px.histogram(practices_list)
fig.add_vline(x=mean_days, line_dash = 'dash', line_color = 'firebrick')
fig.add_vline(x=median_days, line_dash = 'dash', line_color = 'green')

fig.show()

## Your mini-report:

Notice, this is not a typical 'code' cell! It is called 'Markdown' and you can change a type of cell in the meny next to the Run button in the top menu.

**Markdown Cells** are text. you can highlight things by surrpounfing them with ** and you can make headers by starting lines with # or ##

**Code Cells** are just normal cells with python code you are used to

Write your report in a Markdown cell like this one

After analysing the number of GP practices which have closed down over the years, we found that on average GP practices were functioning for a total of 8270 days i.e. providing around 22 years of service. The median was 6421 days  